pkg/policy/consumer.go

```diff
@@ -258,37 +245,34 @@ func (c *Consumable) AllowConsumerAndReverseLocked(cache *ConsumableCache, id Nu
 		}).Debug("Allowing reverse direction")
 		if _, ok := reverse.ReverseRules[c.ID]; !ok {
 			reverse.addToMaps(c.ID)
-			reverse.ReverseRules[c.ID] = NewConsumer(c.ID)
+			reverse.ReverseRules[c.ID] = true
 			return true
 		}
 	}
 	log.WithFields(logrus.Fields{
 		logfields.Identity + ".from": c.ID,
 		logfields.Identity + ".to":   id,
-	}).Warn("Allowed a consumer which can't be found in the reverse direction")
+	}).Warn("Allowed an ingress security identity which can't be found in the reverse direction")
 	return changed
 }
 
-// BanConsumerLocked removes the given consumer from the Consumable's consumers
-// map. Must be called with the Consumable mutex locked.
-func (c *Consumable) BanConsumerLocked(id NumericIdentity) {
-	if consumer, ok := c.Consumers[id.StringID()]; ok {
-		log.WithField("consumer", logfields.Repr(consumer)).Debug("Removing consumer")
-		delete(c.Consumers, id.StringID())
+// RemoveIngressIdentityLocked removes the given security identity from Consumable's
+// IngressIdentities map.
+// Must be called with the Consumable mutex locked.
+func (c *Consumable) RemoveIngressIdentityLocked(id NumericIdentity) {
+	if _, ok := c.IngressIdentities[id]; ok {
+		log.WithField(logfields.Identity, id).Debug("Removing ingress identity")
+		delete(c.IngressIdentities, id)
 
 		if c.wasLastRule(id) {
 			c.removeFromMaps(id)
 		}
-
-		if consumer.Reverse != nil {
-			c.deleteReverseRule(id, c.ID)
-		}
```

Comment:
How is the removal of the reverse rule done now?

Reply:
The Consumable has a separate map, `ReverseRules`.

Reply:
But now we don't delete the (reverse) rule from that map like we did here, by calling `c.deleteReverseRule()`. We still add entries to that map like before (in `AllowConsumerAndReverseLocked()`) , though. IMO we should still be able to track if a reverse entry was added and then remove it here correspondingly.

Reply:
This code was never invoked. The `Reverse` field of any Consumer was never populated with anything, and thus this codepath was never reached. That was part of my rationale for removing the Consumer type all-together, because most of its fields were not used (Decision and Reverse) or Redundant (Identity).

Reply:
>This code was never invoked.

which code? the reverse map was created here:

```
// allowConsumer must be called with global endpoint.Mutex held
func (e *Endpoint) allowConsumer(owner Owner, id policy.NumericIdentity) bool {
	cache := policy.GetConsumableCache()
	if !e.Opts.IsEnabled(OptionConntrack) {
		return e.Consumable.AllowConsumerAndReverseLocked(cache, id)
	}
	return e.Consumable.AllowConsumerLocked(cache, id)
}
```

Reply:
@aanm by 'this code' I meant the code within the if-condition:
```
if consumer.Reverse != nil {
	c.deleteReverseRule(id, c.ID)
}
```

What I meant was that the `Reverse` field of a Consumer was never non-nil.

Reply:
OK, so even when an entry was added to ReverseRules, the `Reverse` field of a consumer was never set to indicate it? So this has never worked as intended, meaning that the `BanConsumerLocked()` never reversed the changes of `AllowConsumerAndReverseLocked()`? If so you should at least leave a TODO etc. to indicate that.

Reply:
OK, will add a TODO.